toolchain/check/generic.cpp

```diff
@@ -23,6 +26,124 @@ auto StartGenericDefinition(Context& context) -> void {
+// current instruction block. Returns a symbolic type ID that refers to the
+// substituted type in each instance of the generic.
+static auto AddGenericTypeToEvalBlock(
+    Context& context, SemIR::GenericId generic_id,
+    SemIR::GenericInstIndex::Region region,
+    Map<SemIR::InstId, SemIR::InstId>& constants_in_generic,
+    SemIR::TypeId type_id) -> SemIR::TypeId {
+  // Check for instructions we've already substituted and return the known
+  // value.
+  auto subst_fn = [&](SemIR::InstId& inst_id) -> bool {
+    if (context.constant_values().Get(inst_id).is_template()) {
+      // This instruction is a template constant, so can't contain any
+      // bindings that need to be substituted.
+      return true;
+    }
+
+    // If this instruction is in the map, return the known result.
+    if (auto result = constants_in_generic.Lookup(inst_id)) {
+      inst_id = result.value();
+      CARBON_CHECK(inst_id.is_valid());
+      return true;
+    }
+    return false;
+  };
+
+  // Build a new instruction in the eval block corresponding to the given
+  // constant.
+  auto rebuild_fn = [&](SemIR::InstId orig_inst_id,
+                        SemIR::Inst new_inst) -> SemIR::InstId {
+    // TODO: Add a function on `Context` to add the instruction without
+    // inserting it into the dependent instructions list or computing a constant
+    // value for it.
+    auto inst_id = context.sem_ir().insts().AddInNoBlock(
+        SemIR::LocIdAndInst::NoLoc(new_inst));
+    auto result = constants_in_generic.Insert(orig_inst_id, inst_id);
+    CARBON_CHECK(result.is_inserted())
+        << "Substituted into an instruction that was already in the map.";
+    auto const_id = AddGenericConstantToEvalBlock(context, generic_id, region,
+                                                  orig_inst_id, inst_id);
+    context.constant_values().Set(inst_id, const_id);
+    return inst_id;
+  };
+
+  // Substitute into the type's constant instruction and rebuild it in the eval
+  // block.
+  auto type_inst_id = SubstInst(context, context.types().GetInstId(type_id),
+                                subst_fn, rebuild_fn);
+  return context.GetTypeIdForTypeInst(type_inst_id);
+}
+
+// Builds and returns a block of instructions whose constant values need to be
+// evaluated in order to resolve a generic instance.
+static auto MakeGenericEvalBlock(Context& context, SemIR::GenericId generic_id,
+                                 SemIR::GenericInstIndex::Region region)
+    -> SemIR::InstBlockId {
+  context.inst_block_stack().Push();
+
+  Map<SemIR::InstId, SemIR::InstId> constants_in_generic;
+  // TODO: For the definition region, populate constants from the declaration.
+
+  // TODO: See if we can ensure that the generic region stack is unchanged
+  // throughout this work. If so, we can use a range-based loop here instead.
+  for (size_t i = 0;
+       i != context.generic_region_stack().PeekDependentInsts().size(); ++i) {
```

Comment:
```suggestion
  for (auto i : llvm::seq(context.generic_region_stack().PeekDependentInsts().size())) {
```

Reply:
Hm, yeah, while the collection could get reallocated (eg, by import ref handling), the number of instructions at this level on the stack really shouldn't change. Made this change and also added a CHECK.